### PR TITLE
[WIP] ZIP 212

### PR DESCRIPTION
**Pull request by therealyingtong**
Draft PR. Partially implements [ZIP 212](https://github.com/zcash/zips/pull/222), addresses https://github.com/zcash/zcash/issues/4557. 

- defines new note encryption `SaplingNoteEncryption::new_v2` which generates `rcm`, `esk` from `rseed`
- defines new note plaintext encryption  `SaplingNoteEncryption::encrypt_note_plaintext_v2` with version `0x02` version and replaces `rcm` with `rseed`
- adds logic for decrypting `v2` plaintext in `parse_note_plaintext_without_memo`
- introduce grace period for `v1` notes received past Canopy activation height
- update old tests 

TODO:
- add new tests
- should mirror C++ implementation https://github.com/zcash/zcash/pull/4578.

### CODE REVIEW

File: zcash_primitives/src/note_encryption.rs [hunk 1]
```diff
@@ -337,31 +395,51 @@ impl SaplingNoteEncryption {
 
 fn parse_note_plaintext_without_memo(
     ivk: &Fs,
+    epk: &edwards::Point<Bls12, PrimeOrder>,
```
Comment (therealyingtong) [inline zcash_primitives/src/note_encryption.rs:398]: pass in optional `esk`, to be used when called by sender (i.e. when using outgoing viewing key)

File: zcash_primitives/src/note_encryption.rs [hunk 2]
```diff
@@ -438,7 +516,7 @@ pub fn try_sapling_compact_note_decryption(
     plaintext.copy_from_slice(&enc_ciphertext);
     ChaCha20Ietf::xor(key.as_bytes(), &[0u8; 12], 1, &mut plaintext);
 
-    parse_note_plaintext_without_memo(ivk, cmu, &plaintext)
+    parse_note_plaintext_without_memo(ivk, epk, cmu, &plaintext)
 }
 
 /// Recovery of the full note plaintext by the sender.
```
Comment (therealyingtong) [inline zcash_primitives/src/note_encryption.rs:522]: this code is repeated in `parse_note_plaintext_without_memo`. try to abstract it into a separate method.

File: zcash_primitives/src/note_encryption.rs [hunk 3]
```diff
@@ -250,18 +257,41 @@ pub struct SaplingNoteEncryption {
...
     memo: Memo,
     ovk: OutgoingViewingKey,
+    rseed: Option<[u8; 32]>,
 }
 
 impl SaplingNoteEncryption {
-    /// Creates a new encryption context for the given note.
-    pub fn new<R: RngCore + CryptoRng>(
+    /// Creates a new encryption context for the given note (before Canopy).
+    pub fn new_v1<R: RngCore + CryptoRng>(
```
Comment (therealyingtong) [inline zcash_primitives/src/note_encryption.rs:265]: we want `new_v1()` to be deterministic. define `generate_v1(rng: &mut R)` which calls `new_v1()`. similarly for `generate_v2()`, `new_v2()`.

File: zcash_primitives/src/transaction/builder.rs [hunk 4]
```diff
@@ -56,15 +57,17 @@ struct SpendDescriptionInfo {
     merkle_path: MerklePath<Node>,
 }
 
-pub struct SaplingOutput {
```
Comment (therealyingtong) [inline zcash_primitives/src/transaction/builder.rs:59]: revert this change.

File: zcash_primitives/src/transaction/builder.rs [hunk 5]
```diff
@@ -93,23 +109,41 @@ impl SaplingOutput {
...
-        );
+        let encryptor;
 
+        if height
+            < <consensus::MainNetwork as consensus::Parameters>::activation_height(
+                consensus::NetworkUpgrade::Canopy,
+            )
+            .unwrap()
+        {
+            encryptor = SaplingNoteEncryption::new_v1(
```
Comment (therealyingtong) [inline zcash_primitives/src/transaction/builder.rs:131]: call `generate_v1()` instead

File: zcash_primitives/src/transaction/builder.rs [hunk 6]
```diff
@@ -93,23 +109,41 @@ impl SaplingOutput {
...
-        let encryptor = SaplingNoteEncryption::new(
-            self.ovk,
-            self.note.clone(),
-            self.to.clone(),
-            self.memo,
-            rng,
-        );
+        let encryptor;
 
+        if height
```
Comment (therealyingtong) [inline zcash_primitives/src/transaction/builder.rs:125]: add in `ZIP212_GRACE_PERIOD`

File: zcash_primitives/src/transaction/builder.rs [hunk 7]
```diff
@@ -93,23 +109,41 @@ impl SaplingOutput {
...
+        {
+            encryptor = SaplingNoteEncryption::new_v1(
+                self.ovk,
+                self.note.clone(),
+                self.to.clone(),
+                self.memo,
+                rng,
+            );
+        } else {
+            encryptor = SaplingNoteEncryption::new_v2(
```
Comment (therealyingtong) [inline zcash_primitives/src/transaction/builder.rs:139]: call `generate_v2` instead

File: zcash_primitives/src/transaction/builder.rs [hunk 8]
```diff
@@ -120,7 +154,19 @@ impl SaplingOutput {
 
         let cmu = self.note.cm(&JUBJUB);
 
-        let enc_ciphertext = encryptor.encrypt_note_plaintext();
+        let enc_ciphertext;
+
+        if height
```
Comment (therealyingtong) [inline zcash_primitives/src/transaction/builder.rs:159]: add in `ZIP212_GRACE_PERIOD`

File: zcash_primitives/src/note_encryption.rs [hunk 9]
```diff
@@ -337,31 +395,51 @@ impl SaplingNoteEncryption {
...
-            .expect("slice is the correct length"),
-    ))?;
-
     let diversifier = Diversifier(d);
     let pk_d = diversifier
         .g_d::<Bls12>(&JUBJUB)?
         .mul(ivk.to_repr(), &JUBJUB);
 
+    let rcm;
+    if plaintext[0] == 0x01 {
```
Comment (therealyingtong) [inline zcash_primitives/src/note_encryption.rs:419]: check if lead byte is consistent with tx blockheight

File: zcash_primitives/src/note_encryption.rs [hunk 10]
```diff
@@ -337,31 +395,51 @@ impl SaplingNoteEncryption {
 
 fn parse_note_plaintext_without_memo(
```
Comment (therealyingtong) [inline zcash_primitives/src/note_encryption.rs:396]: take in tx blockheight as an argument

File: zcash_primitives/src/transaction/builder.rs [hunk 11]
```diff
@@ -79,7 +82,20 @@ impl SaplingOutput {
             return Err(Error::InvalidAmount);
         }
 
-        let rcm = Fs::random(rng);
+        let rcm;
+        let mut rseed = [0u8; 32];
+        if height
```
Comment (therealyingtong) [inline zcash_primitives/src/transaction/builder.rs:87]: add in `ZIP212_GRACE_PERIOD`

File: zcash_primitives/src/transaction/builder.rs [hunk 12]
```diff
@@ -377,7 +423,8 @@ impl<R: RngCore + CryptoRng> Builder<R> {
         value: Amount,
         memo: Option<Memo>,
     ) -> Result<(), Error> {
-        let output = SaplingOutput::new(&mut self.rng, ovk, to, value, memo)?;
+        let output =
+            SaplingOutput::new(self.mtx.expiry_height, &mut self.rng, ovk, to, value, memo)?;
```
Comment (therealyingtong) [inline zcash_primitives/src/transaction/builder.rs:427]: (look at line 355)
define `height` field in `Builder` struct, and compute `expiry_height` in `Builder`

File: zcash_primitives/src/transaction/builder.rs [hunk 13]
```diff
@@ -588,18 +636,35 @@ impl<R: RngCore + CryptoRng> Builder<R> {
...
+                            consensus::NetworkUpgrade::Canopy,
+                        )
+                        .unwrap()
+                    {
+                        rcm = Fs::random(&mut self.rng);
+                        esk = generate_esk_v1(&mut self.rng);
+                    } else {
+                        // rseed is a random 32-byte sequence
+                        let mut rseed = [0u8; 32];
+                        self.rng.fill_bytes(&mut rseed);
```
Comment (therealyingtong) [inline zcash_primitives/src/transaction/builder.rs:652]: this would go into `generate_v2`

File: zcash_primitives/src/note_encryption.rs [hunk 14]
```diff
@@ -133,7 +136,7 @@ impl str::FromStr for Memo {
     }
 }
 
-pub fn generate_esk<R: RngCore + CryptoRng>(rng: &mut R) -> Fs {
+pub fn generate_esk_v1<R: RngCore + CryptoRng>(rng: &mut R) -> Fs {
```
Comment (therealyingtong) [inline zcash_primitives/src/note_encryption.rs:139]: `generate_v1` should generate `esk`

File: zcash_primitives/src/note_encryption.rs [hunk 15]
```diff
@@ -142,6 +145,10 @@ pub fn generate_esk<R: RngCore + CryptoRng>(rng: &mut R) -> Fs {
     Fs::to_uniform(&buffer[..])
 }
 
+pub fn generate_esk_v2(rseed: [u8; 32]) -> Fs {
```
Comment (therealyingtong) [inline zcash_primitives/src/note_encryption.rs:148]: `generate_v2` should generate `rseed`, from which `rcm`, `esk` are derived

File: zcash_primitives/src/note_encryption.rs [hunk 16]
```diff
@@ -250,18 +260,19 @@ pub struct SaplingNoteEncryption {
...
     memo: Memo,
     ovk: OutgoingViewingKey,
+    rseed: Option<[u8; 32]>,
 }
 
 impl SaplingNoteEncryption {
     /// Creates a new encryption context for the given note.
-    pub fn new<R: RngCore + CryptoRng>(
+    pub fn new(
+        esk: Fs,
```
Comment (therealyingtong) [inline zcash_primitives/src/note_encryption.rs:269]: check that `esk`, `rcm` are consistent with `rseed`

File: zcash_primitives/src/note_encryption.rs [hunk 17]
```diff
@@ -284,8 +296,8 @@ impl SaplingNoteEncryption {
         &self.epk
     }
 
-    /// Generates `encCiphertext` for this note.
-    pub fn encrypt_note_plaintext(&self) -> [u8; ENC_CIPHERTEXT_SIZE] {
+    /// Generates `encCiphertext` for v1 note.
+    pub fn encrypt_note_plaintext_v1(&self) -> [u8; ENC_CIPHERTEXT_SIZE] {
```
Comment (therealyingtong) [inline zcash_primitives/src/note_encryption.rs:300]: instead of defining `v1`, v2`, just check `rseed.isSome()`

File: zcash_primitives/src/note_encryption.rs [hunk 18]
```diff
@@ -335,33 +374,76 @@ impl SaplingNoteEncryption {
     }
 }
 
+fn check_note_plaintext_version(height: u32, lead_byte: u8) -> Option<()> {
```
Comment (therealyingtong) [inline zcash_primitives/src/note_encryption.rs:377]: we can make `0x02` valid even before Canopy activation height

File: zcash_primitives/src/note_encryption.rs [hunk 19]
```diff
@@ -496,22 +581,31 @@ pub fn try_sapling_output_recovery(
...
+            plaintext[20..COMPACT_NOTE_SIZE]
+                .try_into()
+                .expect("slice is the correct length"),
+        ))?;
+    } else if plaintext[0] == 0x02 {
+        let mut rseed = [0u8; 32];
+        rseed.copy_from_slice(&plaintext[20..COMPACT_NOTE_SIZE]);
+        rcm = Fs::to_uniform(prf_expand(&rseed, &[0x04]).as_bytes());
+        let derived_esk = Fs::to_uniform(prf_expand(&rseed, &[0x05]).as_bytes());
+        assert_eq!(esk, derived_esk);
```
Comment (therealyingtong) [inline zcash_primitives/src/note_encryption.rs:605]: instead of `assert_eq` use `return None` to avoid crashing the client

File: zcash_primitives/src/note_encryption.rs [hunk 20]
```diff
@@ -690,20 +788,23 @@ mod tests {
...
         let ivk = Fs::random(&mut rng);
 
         let (ovk, ivk, cv, cmu, epk, enc_ciphertext, out_ciphertext) =
-            random_enc_ciphertext_with(ivk, rng);
+            random_enc_ciphertext_with_v1(ivk, rng);
 
-        assert!(try_sapling_note_decryption(&ivk, &epk, &cmu, &enc_ciphertext).is_some());
+        assert!(try_sapling_note_decryption(height, &ivk, &epk, &cmu, &enc_ciphertext).is_some());
         assert!(try_sapling_compact_note_decryption(
+            height,
```
Comment (therealyingtong) [inline zcash_primitives/src/note_encryption.rs:799]: some changes to the c++ code has to be made to pass in `height`

File: zcash_primitives/src/transaction/builder.rs [hunk 21]
```diff
@@ -282,6 +323,7 @@ impl TransactionMetadata {
 /// Generates a [`Transaction`] from its inputs and outputs.
 pub struct Builder<R: RngCore + CryptoRng> {
     rng: R,
+    height: Option<u32>,
```
Comment (therealyingtong) [inline zcash_primitives/src/transaction/builder.rs:326]: `height` should not be optional

File: zcash_primitives/src/transaction/builder.rs [hunk 22]
```diff
@@ -561,8 +605,9 @@ impl<R: RngCore + CryptoRng> Builder<R> {
                 // Record the post-randomized output location
                 tx_metadata.output_indices[pos] = i;
 
-                output.build(prover, &mut ctx, &mut self.rng)
+                output.build(self.mtx.expiry_height, prover, &mut ctx, &mut self.rng)
```
Comment (nuttycom) [inline zcash_primitives/src/transaction/builder.rs:608]: This should probably be the current height, rather than the expiry height?

**Comment by therealyingtong** [comment]
Superseded by https://github.com/zcash/librustzcash/pull/258 in the interest of mirroring C++ implementation
